dali/test/python/test_operator_noise_salt_and_pepper.py

```diff
@@ -0,0 +1,131 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from nvidia.dali import pipeline_def
+from nvidia.dali.backend_impl import TensorListGPU
+import nvidia.dali.ops as ops
+import nvidia.dali.fn as fn
+import nvidia.dali.types as types
+import random
+import numpy as np
+import os
+from test_utils import get_dali_extra_path, check_batch
+from test_noise_utils import PSNR
+
+test_data_root = get_dali_extra_path()
+images_dir = os.path.join(test_data_root, 'db', 'single', 'png')
+dump_images = False
+
+def salt_and_pepper_noise_ref(x, prob, salt_to_pepper_prob, per_channel, salt_val, pepper_val):
+    x = np.array(x, dtype=np.float32)
+    salt_prob = prob * salt_to_pepper_prob
+    pepper_prob = prob * (1.0 - salt_to_pepper_prob)
+    nchannels = x.shape[-1]
+    mask_sh = x.shape if per_channel else x.shape[:-1]
+    mask = np.random.choice([pepper_val, np.nan, salt_val], p=[pepper_prob, 1 - prob, salt_prob], size=mask_sh)
+    if not per_channel:
+        mask = np.stack([mask] * nchannels, axis=-1)
+    y = np.where(np.isnan(mask), x, mask).astype(np.uint8)
+    return y
+
+@pipeline_def
+def pipe_salt_and_pepper_noise(prob, salt_to_pepper_prob, channel_first, per_channel, salt_val, pepper_val, device='cpu'):
+    encoded, _ = fn.readers.file(file_root=images_dir)
+    in_data = fn.decoders.image(encoded, output_type=types.RGB)
+    if device == 'gpu':
+        in_data = in_data.gpu()
+    if channel_first:
+        in_data = fn.transpose(in_data, perm=[2, 0, 1])
+    prob_arg = prob or fn.random.uniform(range=(0.05, 0.5))
+    salt_to_pepper_prob_arg = salt_to_pepper_prob or fn.random.uniform(range=(0.25, 0.75))
+    out_data = fn.noise.salt_and_pepper(
+        in_data, per_channel=per_channel, prob=prob_arg, salt_to_pepper_prob=salt_to_pepper_prob_arg,
+        salt_val=salt_val, pepper_val=pepper_val
+    )
+    return in_data, out_data, prob_arg, salt_to_pepper_prob_arg
+
+def verify_salt_and_pepper(output, input, prob, salt_to_pepper_prob, per_channel, salt_val, pepper_val):
+    assert output.shape == input.shape
+    height, width, nchannels = output.shape
+    npixels = height * width
+    salt_count = 0
+    pepper_count = 0
+    pixel_count = 0
+    if per_channel:
+        output = np.reshape(output, (npixels * nchannels, 1))
+        input = np.reshape(input, (npixels * nchannels, 1))
+    passthrough_mask = np.all(output == input, axis=-1)
+    pepper_mask = np.all(output == pepper_val, axis=-1)
+    salt_mask = np.all(output == salt_val, axis=-1)
+    pixel_mask = np.logical_and(np.all(input != pepper_val, axis=-1),
```

Comment:
How it is different from passthrough_mask?

Reply:
Now I see. How about renaming it to `input_pixel_mask`?

Reply:
I would add a comment that you want to filter out pixels in the input that already have salt/pepper values.

Reply:
Will do